src/number_theory/modular_forms/modular_forms.lean

```diff
@@ -0,0 +1,391 @@
+@[simp]
+lemma wmodular_mem' (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) :
+  f ∈ (weakly_modular_submodule k Γ) ↔ ∀ γ : Γ, ∀ z : ℍ,
+  f (γ • z) = ((↑ₘγ 1 0 : ℝ) * z +(↑ₘγ 1 1 : ℝ))^k * f z :=
+begin
+  simp only [wmodular_mem],
+  split,
+  { intros h1 γ z,
+  have h3 : (f ∣[k] γ) z = f z , by {simp_rw (h1 γ)},
+  rw [←h3, slash_k, mul_comm],
+  have h55 := zpow_neg_mul_zpow_self k (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) z),
+  simp only [upper_half_plane.denom, upper_half_plane.subgroup_to_sl_moeb, upper_half_plane.sl_moeb,
+    coe_coe, matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast] at *,
+  rw [mul_assoc, h55, ←int.coe_cast_ring_hom, ←matrix.special_linear_group.coe_matrix_coe,
+    matrix.special_linear_group.det_coe ((γ : SL(2, ℤ)) : SL(2, ℝ))],
+  simp only [of_real_one, one_zpow, mul_one], },
+  { intros hf γ,
+  simp_rw slash_k,
+  ext1,
+  rw [←upper_half_plane.subgroup_moeb, (hf γ x), mul_comm],
+  have h55 := zpow_neg_mul_zpow_self k (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) x),
+  simp_rw upper_half_plane.denom at *,
+  simp only [matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix, coe_coe,
+    matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast] at h55,
+  simp only [coe_coe, matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.map_apply, of_real_int_cast],
+  rw (matrix.special_linear_group.det_coe ((γ : SL(2, ℤ)) : SL(2, ℝ))),
+  simp only [matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast, of_real_one, one_zpow, mul_one],
+  simp_rw [← mul_assoc, h55],
+  simp },
+end
+
+lemma mul_modular (k_1 k_2 : ℤ) (Γ : subgroup SL(2,ℤ)) (f g : ℍ → ℂ)
+  (hf : f ∈ weakly_modular_submodule k_1 Γ) (hg : g ∈ weakly_modular_submodule k_2 Γ) :
+  f * g ∈ weakly_modular_submodule (k_1 + k_2) Γ :=
+begin
+  simp only [wmodular_mem', pi.mul_apply, coe_coe] at *,
+  intros γ z,
+  rw [(hf γ z), (hg γ z)],
+  have pown := zpow_add₀ (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) z) k_1 k_2,
+  simp only [upper_half_plane.denom, coe_fn_coe_base, ne.def,
+    matrix.general_linear_group.coe_fn_eq_coe, coe_coe] at pown,
+  rw pown,
+  ring,
+end
+
+/--Filter for approaching `i∞`-/
+def at_I_inf := filter.at_top.comap upper_half_plane.im
+
+lemma at_I_inf_mem (S : set ℍ) : S ∈ at_I_inf ↔ (∃ A : ℝ, ∀ z : ℍ, A ≤ im z → z ∈ S) :=
+begin
+  simp only [at_I_inf, filter.mem_comap', filter.mem_at_top_sets, ge_iff_le, set.mem_set_of_eq,
+    upper_half_plane.coe_im],
+  split,
+  { intro h, cases h with a h, refine ⟨a, (λ z hz, by {apply h (im z) hz , refl})⟩ },
+  { refine (λ h, by {cases h with A h,
+    refine ⟨A, (λ b hb x hx, by {apply (h x), rw hx, exact hb})⟩}) }
+end
+
+/--A function ` f : ℍ → ℂ` is bounded at infinity if there exist real numbers `M,A` such that
```

Comment:
This notion is interesting regardless of modular forms, can you move it to another file? (Maybe a new file.) You can even open another PR for this.. it will probably speed up the review process.

Reply:
Sure I've made a new file for this. I wanted to keep it all together so the that whole defn of modular forms would get PRed at once, but I can split the PR if you think its better. Also, any preference on where to put this file? I just put it in the modular forms folder for the moment.

Reply:
I think it can go in the `upper_half_plane` folder.